text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,264 @@
+
+// Multiple hints
+import("c-bar-mobile" /* "@salesforce/client/formFactor": "SMALL && LARGE"*/);
+import("c-bar-mobile" /* "@salesforce/client/formFactor": "SMALL" && "@salesforce/client/formFactor": "LARGE"*/);
+
+// Dynamic module name specifier
+const bar = 'module-name';
+import(bar/* "@salesforce/client/formFactor": "SMALL" */)
+```
+
+#### Separation of Concern
+Dynamic imports hint metadata gathering happens during component compilation. An implicit invariant of this process is that the metadata collection will **not** fail due to a bad hint. This could be a hint not following specified grammar or not using a well understood hint value. Validation of the hint keys and values will not be the concern of this RFC.
+
+Additional validation and post processing of the gathered metadata is not a concern of this RFC. That logic will be a responsibility of the business logic that consumes the metadata.
+
+Similarly, any design time validations of the hint grammar should be handled by linting the code. For example, on the salesforce platform, the allowed hint keys might be standardized to only `@salesforce` scoped specifiers. The values will be validated based on the relevant value for the specifier. An ESLint rule will enforce this restriction via the [eslint-plugin-lwc](https://github.com/salesforce/eslint-plugin-lwc).
+
+### Reasons
+The hint syntax in the form of a comment was selected for the following reasons:
+* Complies with ‘what you write is what you see’ concept - an originally authored code remains untouched
+* Eliminates any additional code transformations at compile time
+* Eliminates any additional processing during module resolution at runtime (to strip out or pre-process module name)
+
+### Dynamic Import Metadata
+The proposed metadata format will include dynamically imported component name and a hint object, which in-itself contains the original query and a hint value. A hint value will be an array of objects with raw hint query, expected hint condition, and hint resource information.
+
+Proposed Shape
+```ts
+export interface Range {
+    start: number;
+    length: number;
+}
+
+export interface DynamicImportHint {
+    rawValue: string; // Trimmed hint phrase
+    specifier: {
+        id: string;
+        namespacedId: string | undefined;
+        type: string | undefined; // Used only if the hint key is a well known reference type
+        file: string;
+        locations: Range[];
+    };
+    value: string;
+}
+
+export interface DynamicImportMetadata {
+    moduleName: string;
+    file: string;
+    hints: DynamicImportHint[];
+}
+```
+
+Example:
+```js
+{
+    dynamicImports: [{
+        moduleName: 'lightning-foo',
+        file: 'foo.js'
+        hints: [{
+            rawValue: '/* "@salesforce/client/formFactor": "SMALL" */'
+            specifier: {
+                id: '@salesforce/client/formFactor',
+                namespacedId: undefined, // scoped modules are namespaced at compile time
```

Comment:
The hint key is not always a module, right? It could be anything? The compiler metadata probably shouldn't dictate your hint's meaning?

Reply:
I echo this as well, the hints should just be a key/value pair. I don't see the value of restricting keys to valid modules. For example, we might add in the future: 
- a `prefetch` hint that will always prefetch the chunk
- or a `chunkName` like in webpack to force the name of the generated chunk.

Reply:
The intention was to reuse the [Reference](https://git.soma.salesforce.com/lwc/lwc-api/blob/master/src/main/java/com/salesforce/lwc/reference/Reference.java#L16) data structure. I am fine to open this up to a more generic definition like
```
 interface DynamicImportHint {
    rawValue: string;
    hintKey: string;
    hintValue: string;
    location: SourceLocation;
}  
```
_SourceLocation: From @babel/types_

@apapko fyi